toolchains/python/python.bzl

```diff
@@ -200,3 +200,184 @@ def nixpkgs_python_configure(
+                dep[PyInfo].transitive_sources
+                for dep in ctx.attr.deps
+            ]),
+        ),
+    ]
+
+python_module = rule(
+    implementation = _python_module_impl,
+    attrs = {
+        "store_path": attr.label(
+            allow_single_file = True,
+            doc = "nix store path of python package",
+        ),
+        "files": attr.label_list(
+            allow_files = True,
+        ),
+        "deps": attr.label_list(
+            providers = [PyInfo],
+        ),
+    },
+    executable = False,
+    test = False,
+)
+""",
+    )
+
+    # 5. generate //:requirements.bzl for later import.
+    repository_ctx.file("requirements.bzl", """
+def requirement(package_name):
+    return "@{}//:{{}}".format(package_name)
+""".format(repository_ctx.name),
+    )
+
+    # TODO: make it lazy in the packages themselves ?
+
+
+_nixpkgs_python_repository = repository_rule(
+    _nixpkgs_python_repository_impl,
+    attrs = {
+        "json_deps": attr.label(),
+    },
+)
+
+
+def nixpkgs_python_repository(
+        name,
+        repository = None,
+        repositories = {},
+        nix_file = None,
+        nix_file_deps = [],
+        quiet = False,
+        ):
+    """Define a collection of python modules.
+
+    Args:
+      name: The name for the created module set.
+      repository: See [`nixpkgs_package`](#nixpkgs_package-repository).
+      repositories: See [`nixpkgs_package`](#nixpkgs_package-repositories).
+      nix_file: See [`nixpkgs_package`](#nixpkgs_package-nix_file).
+      nix_file_deps: See [`nixpkgs_package`](#nixpkgs_package-nix_file_deps).
+    """
+
+    generated_deps_name = "_generated_{}_deps".format(name)
+
+    nixpkgs_package(
+        name = generated_deps_name,
+        nix_file_content = """
+{ nix_file }:
+let
+  nixpkgs = import <nixpkgs> {};
+  pythonExpr = import nix_file;
+  inherit (pythonExpr) python pkgs;
+
+  isPythonModule = drv: drv ? pythonModule && drv ? pythonPath;
+  filterPythonModules = builtins.filter isPythonModule;
+
+  # Ensure the dependency list is unique, otherwise bazel complains about
+  # duplicate names in the generated python_module() rule
+  unique = list: builtins.attrNames (builtins.listToAttrs (builtins.map (x: {
+    name = x;
+    value = null;
+  }) list));
+
+  # Build the list of python modules from the initial set in `pkgs`.
+  # Each key is the package name, and the value is the derivation itself.
+  toClosureFormat = builtins.map (drv: {
+    key = drv.pname;
+    value = drv // { _pythonModules = filterPythonModules drv.propagatedBuildInputs; };
+  });
+  startSet = toClosureFormat pkgs;
+  closure = builtins.genericClosure {
+    inherit startSet;
+    operator = item: toClosureFormat (filterPythonModules item.value.propagatedBuildInputs);
+  };
+
+  # Using the information generated above, map the package information into
+  # a list, described in the Output description at the top of this file.
```

Comment:
Which "Output description" does this refer to?

Reply:
Leftover from [redacted] :-D.